kpet/data.py

```diff
@@ -451,6 +474,54 @@ def convert_tree_arches(self):
 
             self.trees[name]["arches"] = list(tree_arches)
 
+    def convert_sets(self):
+        """
+        Convert each suite / cases sets from a list of regexes
+        to a list of set names matching those regexes
+
+        Raises:
+            A schema.Invalid exception when finding an invalid regex
+        """
+
+        error = ("One of '{0}: set:' regexes\n" +
+                 "doesn't match any of the available sets\n" +
+                 "The regex: '{1}'\n" +
+                 "The avaliable sets: " + str(list(self.sets.keys())))
+
+        # TODO: Remove after conversion
+        # This function is to convert from the old format (where sets were a
+        # part of the pattern)
+        def convert_pattern(pattern):
+            matches = set()
+            for set_name in self.sets:
+                target = Target(trees=Target.ANY,
+                                arches=Target.ANY,
+                                components=Target.ANY,
+                                sets=set(set_name),
+                                sources=Target.ANY)
+                if pattern.matches(target):
```

Comment:
Ran into issues here because sometimes the whole pattern is at the suite while the case itself has no pattern on it's own, this means the pattern.data of that case is an empty dict and that passes as "matching" every time here

Reply:
Adding a check for an empty patterns seems to fix most situations but I don't think that's a good solution or that it fixes the whole issue

Reply:
Yeah the empty pattern check will help in most situations but doesn't cover the situation where the pattern doesn't have a "sets" property which happens in 2 scenarios:
- The case/suite is unassigned
- The case/suite already converted to having "sets" outside the pattern
Checking for the second situation would be easy and I suppose we could call the __node_matches function directly somehow to only check for the sets but that would be ugly.

Also I feel like a function to extract a single qualifier from a pattern could be usefull in other situations as well so that might be the way to go

Reply:
Maybe just a check if a pattern "cares" about a qualifier whatsoever (if the qualifier name apears anywhere in the pattern "tree"?) would be enough

Reply:
First, I think you could make it to only call `convert_pattern()` if `suite.sets`/`case.sets` is unset. I.e. if there are no new-style sets.

Regarding the matching, I think we could say that if a pattern matches every possible set name, that means it has no restrictions, which means we didn't have membership defined at that level. Therefore we don't need to add the "sets" property for that suite/case. So just check that the filtered set of set names is the same as the total set of set names, and don't add the "sets" property in that case.

I'll need to think more about the matching logic around this, though, and I'm all out of my thinking capacity with the cold right now. Hopefully I'll get to this soon.